### DISCUSSION TITLE
URL defaults are not respected.

### DISCUSSION THREAD

**Discussion by Mecit** [livewire/livewire#2473] [1. Help] [answered]
Hey there. I'm not sure if that's an issue or I have some conceptual misunderstanding, so I'm starting this as a discussion. 

I have a subdomain-based tenancy application and it uses a route level middleware that sets the default value for tenants. 

```
Route::middleware(App\Http\Middleware\SetDefaultTenantForRoutes::class)
     ->domain('{tenant}.' . config('app.url'))
    ...
```

And all the middleware does is to set the default tenant with `URL::defaults(['tenant' => $request->route('tenant')])` and that works just fine in a typical Laravel lifecycle. However, it seems I cannot use route redirections or url generations without having to set the tenant explicitly in Livewire components/views. e.g. in a Livewire component method `return redirect()->route('tenant:users')` throws a "missing required parameter" exception. As for views, I can use it with no parameters when a component first renders, but when the component updates, the same error occurs again. Any idea as to why Livewire doesn't pick the parameter up? It'd be quite cumbersome having to set the parameter every time I need to redirect a user or generate a link to a named route.

**Comment by joshhanley** [comment]
@Mecit I'm not sure if that is a Livewire issue or a Laravel thing? Can you test using a normal Laravel controller to see if the redirect is working as expected? Maybe you need to modify the redirect and url helpers to ensure it redirects to the tenants sub-domain, instead of the app url.

If it works for Laravel, then it will be a Livewire issue, and we can dig further to try and work out why. Let us know how you go.

Hope this helps!

**Reply by Mecit** [reply]
@joshhanley yes, it works as expected when I use normal controllers. e.g. The controller responsible for authentication is also registered within the same route group and the parameter is picked up during redirects. For the time being, I tackled this problem by writing a helper function that explicitly passes the tenant parameter to the route.

**Reply by joshhanley** [reply]
@Mecit ok thanks for letting us know. Is there any chance you could put a sample repo together with your middleware and a Livewire component that demonstrates the issue so I can investigate? If you could, that'd be a great help, and hopefully we can get to the bottom of it!

Thanks

**Reply by Mecit** [reply]
Hey, Josh. I set up a quick example demonstrating this problem. Here's the link to the repo: [Mecit/livewire-url-defaults](https://github.com/Mecit/livewire-url-defaults)

 Assuming you're using Valet, after cloning the repo, you can visit http://example.livewire-url-defaults.test/normal-redirect which utilizes a typical Laravel workflow. When you click the button. It will redirect as it should. After that, please, visit http://example.livewire-url-defaults.test/livewire-redirect and click the button. This time you will see an `Illuminate\Routing\Exceptions\UrlGenerationException` exception with the message `Missing required parameter for [Route: tenant:home] [URI: /] [Missing parameter: tenant]` because Livewire doesn't include the said parameter in the headers.

**Comment by A-Lawrence** [comment, marked as answer]
👋🏼 Howdy - Stumbled across this discussion after experiencing the same issue. Adding livewire to a project that otherwise works with URL defaults.

We have a middleware that handles tenant identification and sets the sub-domain slug as a default:

```php
URL::defaults(['organisation' => $organisation->slug]);
````

On the initial page load, this appears to work fine, but any subsequent livewire requests do not seem to respect this default:

```
Missing required parameter for [Route: auth.password.request] [URI: auth/password/reset] [Missing parameter: organisation]
````

**Reply by A-Lawrence** [reply]
**Update** - writing that reply was a great rubber duck...

@Mecit this might also solve your problem.

After doing some digging, I spotted the ["Persistent Middleware" section of the docs](https://laravel-livewire.com/docs/2.x/security). 

For us the solution was as simple as adding this block to the AppServiceProvider:

```php
Livewire::addPersistentMiddleware([
            IdentifyTenant::class,
            CheckOrganisationAccess::class
        ]);
```

**Reply by Mecit** [reply]
@A-Lawrence that's interesting. I also had come across `addPersistentMiddleware` in the docs and when I tried to give it a shot, it kept throwing an error (It was an "undefined method" exception if I don't remember wrong). I tried it again just now and it seems to be working perfectly fine. I just did a composer update last night, maybe that was fixed somewhere between my last two updates. Thanks a lot.

@joshhanley also thank you for your time. The `addPersistentMiddleware` method seems to be working as it should now, so no further action is necessary.

**Reply by A-Lawrence** [reply]
Appears to have only changed 17 days ago: https://github.com/livewire/livewire/commit/bbed161d2f70feae5bc1188b2429d01e3fe820fa

Prior to that, it was defined in a config entry for Livewire ("middlware_group").  Glad it's working for you too!

**Reply by joshhanley** [reply]
Yep persistent middleware was added in 2.3.9 to fix some bugs and for situations like this. Good catch @A-Lawrence I'd forgotten about that and it's a perfect solution to this issue.

For anyone else coming across this:
I'd recommend not using 2.3.9 and make sure you're on the latest version as there was some unintended bugs introduced because of this change.
